### PR TITLE
node graveyard resurrection

### PR DESCRIPTION
**Pull request by dylanlott**
Resurrects nodes from the _graveyard_

Every X interval, the cache checks the Seen() map for nodes that it's talked to and tries to ping them again. If they've been removed from the cache in the last hour for unresponsiveness, this is their chance to be online again and get added back to the cache. If they still don't respond, they stay in Seen() and nothing happens. 

This PR address some comments in #1185

Also improves how we handle tickers in discovery/service.go

### CODE REVIEW

File: pkg/discovery/service.go [hunk 1]
```diff
@@ -144,8 +160,39 @@ func (discovery *Discovery) refresh(ctx context.Context) error {
 	return nil
 }
 
+// graveyard attempts to ping all nodes in the Seen() map from Kademlia and adds them to the cache
+// if they respond. This is an attempt to resurrect nodes that may have gone offline in the last hour
```
Comment (jenlij) [inline pkg/discovery/service.go]: nit: keep the end of the sentence on the same line

File: pkg/discovery/service.go [hunk 2]
```diff
@@ -144,8 +160,39 @@ func (discovery *Discovery) refresh(ctx context.Context) error {
...
+// and were removed from the cache due to an unsuccessful response.
+func (discovery *Discovery) graveyard(ctx context.Context) error {
+	discovery.log.Debug("starting node graveyard resurrection")
+	seen := discovery.kad.Seen()
+	var errors utils.ErrorGroup
+
+	for _, n := range seen {
+		ping, err := discovery.kad.Ping(ctx, *n)
+		if err != nil {
+			discovery.log.Debug("could not ping node in graveyard check")
```
Comment (jenlij) [inline pkg/discovery/service.go]: maybe add the nodeID to the log?

File: pkg/discovery/service.go [hunk 3]
```diff
@@ -144,8 +160,39 @@ func (discovery *Discovery) refresh(ctx context.Context) error {
...
+			// we don't want to report the ping error to ErrorGroup because it's to be expected here.
+			continue
+		}
+
+		err = discovery.cache.Put(ctx, ping.Id, ping)
+		if err != nil {
+			discovery.log.Warn("could not update node uptime")
+			errors.Add(err)
```
Comment (jenlij) [inline pkg/discovery/service.go]: Is there a preference to use errorgroup rather than combineErrors?

**Comment by dylanlott** [comment]
In other places I’ve been discouraged from adding the ID to the logs so 🤷‍♀️
  

  
  

  
  
>

**Comment by jenlij** [comment]
Do we have some unit tests for this?

**Comment by dylanlott** [comment]
Egon wants to move towards using errorgroup, since it’s easier to manage and under the hood it uses combine errors anyway.   
  

  
  

  
  
>

File: pkg/discovery/service.go [hunk 4]
```diff
@@ -144,8 +160,39 @@ func (discovery *Discovery) refresh(ctx context.Context) error {
 	return nil
 }
 
+// graveyard attempts to ping all nodes in the Seen() map from Kademlia and adds them to the cache
+// if they respond. This is an attempt to resurrect nodes that may have gone offline in the last hour
+// and were removed from the cache due to an unsuccessful response.
+func (discovery *Discovery) graveyard(ctx context.Context) error {
+	discovery.log.Debug("starting node graveyard resurrection")
+	seen := discovery.kad.Seen()
```
Comment (aligeti) [inline pkg/discovery/service.go]: how is the graveyard discovery.kad.Seen() different from refresh's discovery.kad.Seen()? Are they looking at the same overlay cache?  graveyard()'s overlay cache(which holds non-pingable nodes) shouldn't it be different overlay cache from the refresh overlay cache?

**Comment by dylanlott** [comment]
The refresh function needs to have the seen logic removed, I’ll do that   
  

  
  

  
  
>

File: pkg/discovery/service.go (hunk 2 above)
Comment (dylanlott) [reply pkg/discovery/service.go]: I've been discouraged from adding node ID's to logs, so I didn't add it here.

File: pkg/discovery/service.go (hunk 3 above)
Comment (dylanlott) [reply pkg/discovery/service.go]: @egonelbre has been pushing for using ErrGroup over CombineErrors. It uses CombineErrors under the hood, anyway.

File: pkg/discovery/service.go [hunk 5]
```diff
@@ -71,24 +74,37 @@ func (discovery *Discovery) Close() error { return nil }
...
-
-		err = discovery.discover(ctx)
-		if err != nil {
-			discovery.log.Error("Error with cache discovery", zap.Error(err))
-		}
-
 		select {
-		case <-ticker.C: // redo
+		case <-refresh.C:
+			discovery.log.Info("REFRESH STARTING")
```
[the remark this answers is not in the document]
Comment (egonelbre) [reply pkg/discovery/service.go]: Yeah, remove the messages, these will start triggering every second for tests and aren't particularly useful.

File: pkg/discovery/service.go [hunk 6]
```diff
@@ -71,24 +74,37 @@ func (discovery *Discovery) Close() error { return nil }
...
+				discovery.log.Error("Error with cache refresh: ", zap.Error(err))
+			}
+		case <-discover.C:
+			discovery.log.Info("DISCOVERY STARTING")
+			err := discovery.discover(ctx)
+			if err != nil {
+				discovery.log.Error("Error with cache discovery: ", zap.Error(err))
+			}
+		case <-graveyard.C:
+			discovery.log.Info("GRAVEYARD RESURRECTION STARTING")
```
Comment (egonelbre) [inline pkg/discovery/service.go]: Remove the message.

File: pkg/discovery/service.go [hunk 7]
```diff
@@ -71,24 +74,37 @@ func (discovery *Discovery) Close() error { return nil }
...
 		select {
-		case <-ticker.C: // redo
+		case <-refresh.C:
+			discovery.log.Info("REFRESH STARTING")
+			err := discovery.refresh(ctx)
+			if err != nil {
+				discovery.log.Error("Error with cache refresh: ", zap.Error(err))
+			}
+		case <-discover.C:
+			discovery.log.Info("DISCOVERY STARTING")
```
Comment (egonelbre) [inline pkg/discovery/service.go]: Remove the messages.

File: pkg/discovery/service.go [hunk 8]
```diff
@@ -71,24 +74,37 @@ func (discovery *Discovery) Close() error { return nil }
...
+				discovery.log.Error("Error with cache discovery: ", zap.Error(err))
+			}
+		case <-graveyard.C:
+			discovery.log.Info("GRAVEYARD RESURRECTION STARTING")
+			err := discovery.graveyard(ctx)
+			if err != nil {
+				discovery.log.Error("graveyard resurrection failed")
+			}
 		case <-ctx.Done():
+			discovery.log.Info("DISCOVERY DONE FIRED")
```
Comment (egonelbre) [inline pkg/discovery/service.go]: Remove message.

File: pkg/discovery/service.go [hunk 9]
```diff
@@ -144,8 +160,39 @@ func (discovery *Discovery) refresh(ctx context.Context) error {
 	return nil
 }
 
+// graveyard attempts to ping all nodes in the Seen() map from Kademlia and adds them to the cache
+// if they respond. This is an attempt to resurrect nodes that may have gone offline in the last hour
+// and were removed from the cache due to an unsuccessful response.
+func (discovery *Discovery) graveyard(ctx context.Context) error {
+	discovery.log.Debug("starting node graveyard resurrection")
```
Comment (egonelbre) [inline pkg/discovery/service.go]: Remove message.

**Comment by egonelbre** [comment]
Is there a test we could do that checks for the intended behavior?

**Pull request merged**